content/blog/z-index-by-examples/index.md

```diff
@@ -0,0 +1,209 @@
+positioned box creates a new **stacking context**.
+
+A stacking what?
+[MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) says:
+
+> The stacking context is a three-dimensional conceptualization of HTML elements along an imaginary z-axis relative to the user, who is assumed to be facing the viewport or the webpage. HTML elements occupy this space in priority order based on element attributes.
+
+I thought this is what we were trying to do all along. Let's keep reading.
+
+The interesting part comes later:
+
+> Within a stacking context, child elements are stacked according to the same rules previously explained. Importantly, the z-index values of its child stacking contexts only have meaning in this parent.
+
+That's the key point. A stacking context will force the `z-index` of its
+child stacking contexts to only have a local meaning. Let's look at our
+mysterious example again:
+
+![mystery part two](./mystery-two.png)
+
+Setting `z-index: 0` on the blue box changed the meaning of the `z-index`
+declaration on the yellow box. Before, the yellow box was out there playing
+with the big boys. Now, it can only affect stacking within the blue box.
+
+This is a great cause of frustration. No matter how high you set a
+`z-index` property, you will _never_ be able to 'escape' the parent
+stacking context.  And every time you are struggling with making `z-index`
+work you can bet that it's because _something_ created a stacking context
+that you can't escape from.
+
+With this newfound understanding, spend some time explaining why [this
+example](https://z-index.juliu.is/?css=b%2Bposition%3A%20relative%3Bz-index%3A%201%7Cg%2Bposition%3A%20relative%3Bz-index%3A%2010%7Cp%2Bposition%3A%20relative%3Bz-index%3A%202%7Cr%2Bposition%3A%20relative%7Cy%2Bposition%3A%20relative%3Bz-index%3A%205%3Bheight%3A%20200px) makes perfect sense:
+
+![no mystery](./no-mystery.png)
+
+## The usual suspects
+
+The
+[MDN article](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context)
+lists all cases when a new stacking context is formed. I'm going to go through a
+list of usual suspects.
+
+[Opacity less than 1](https://z-index.juliu.is/?css=p%2Bopacity%3A%200.90)
+
+![opacity](./opacity.png)
+
+[Transform and filter effects](https://z-index.juliu.is/?css=p%2Btransform%3A%20rotate(30deg))
+
+![transform](./transform.png)
+
+[Flex child with z-index](https://z-index.juliu.is/?css=b%2Bdisplay%3A%20flex%7Cp%2Bposition%3A%20relative%3Bz-index%3A%201%7Cy%2Bz-index%3A%202)
+
+![flex child](./flex-child.png)
+
+[Grid child with z-index](https://z-index.juliu.is/?css=b%2Bdisplay%3A%20grid%7Cg%2Bz-index%3A%202%7Cp%2Bposition%3A%20relative%3Bz-index%3A%201)
+
+![grid child](./grid-child.png)
+
+As before, I recommend going through each example, playing around with
+it, and explaining in your head why it makes perfect sense.
```

Comment:
I've found [this chrome extension](https://chrome.google.com/webstore/detail/z-context/jigamimbjojkdgnlldajknogfgncplbh) very helpful when troubleshooting z-index issues. It gives you an easy way to tell which elements create a stacking context and why:

<img width="1438" alt="Screen Shot 2021-01-27 at 09 57 52" src="https://user-images.githubusercontent.com/753421/105994300-3ec64f00-6086-11eb-9504-9114ed97e35a.png">

Just sharing in case you didn't know of it, or if you think that linking to this or other tools would be appropriate given how hard it is sometimes to troubleshoot these issues on big layouts even if you do know the rules.